fs/src/io_uring/memory.rs

```diff
@@ -14,53 +14,25 @@ use {
 // and chunk it in slices of up to 1G each.
 const FIXED_BUFFER_LEN: usize = 1024 * 1024 * 1024;
 
-pub enum LargeBuffer {
-    Vec(Vec<u8>),
-    HugeTable(PageAlignedMemory),
-}
-
-impl Deref for LargeBuffer {
-    type Target = [u8];
-
-    fn deref(&self) -> &Self::Target {
-        match self {
-            Self::Vec(buf) => buf.as_slice(),
-            Self::HugeTable(mem) => mem.deref(),
+/// Allocate memory buffer optimized for io_uring operations, i.e.
+/// using HugeTable when it is available on the host.
+pub fn new_large_buffer(size: usize) -> std::io::Result<PageAlignedMemory> {
```

Comment:
I would rather we still minimize the size of this PR and not include the mem buffer refactor / change in it. For the Vec case we could simply overallocate it by PAGE_SIZE-1 and then operate on aligned memory skipping the non-aligned address bytes.

The code for doing that could land in this module.

Reply:
I don't know how to use Vec and generate aligned addresses safely, which is why I used mmap. I bashed my head against the wall for a few hours trying to do so without introducing more `unsafe` code and eventually gave up. I would really appreciate your guidance here if possible.

Reply:
I don't think there is anything magical about aligning, those are just pointers that need to be multiples of some constant, so something like that?
```rust
        let buffer = vec.as_mut();
        let align_offset = buffer.as_ptr().align_offset(page_size);
        let aligned_buf = buffer.split_at(align_offset).1;
```

Reply:
I'm looking into using Vec, but an edge case jumped out to me and I'd like to hear your thoughts. What if the Vec allocates mostly inside one page, and then there is only a little allocated in the next page? Then when we split out the aligned buf, it's far smaller than expected. In order for us to avoid this edge case we would have to allocate a buffer of 2*page_size, at which point we might as well just allocate a full page using mmap with far less headache...

Reply:
We should always over-allocate Vec by PAGE_SIZE-1 when direct-io is enabled (i.e. page_aligned: true is passed to LargeBuffer::new) - allocations can often be cheaper than mmap, since they may come from allocator's arena, while mmap will do syscall no matter what.